pyro/infer/mcmc/hmc.py

```diff
@@ -0,0 +1,225 @@
+        detach_nodes(r_next)
+        for site_name in z_next:
+            # r(n+1/2)
+            r_next[site_name] = r_next[site_name] + 0.5 * step_size * (-grads[site_name])
+            # z(n+1)
+            z_next[site_name] = z_next[site_name] + step_size * r_next[site_name]
+        # retain gradients for intermediate nodes in backward step
+        retain_grads(z_next)
+        retain_grads(r_next)
+        grads = grad_potential(z_next)
+        for site_name in r_next:
+            # r(n+1)
+            r_next[site_name] = r_next[site_name] + 0.5 * step_size * (-grads[site_name])
+    return z_next, r_next
+
+
+def retain_grads(z):
+    for key in z:
+        # XXX: can be removed with PyTorch 0.3
+        if z[key].is_leaf and not z[key].requires_grad:
+            z[key].requires_grad = True
+        z[key].retain_grad()
+
+
+def detach_nodes(z):
+    for key in z:
+        z[key] = Variable(z[key].data, requires_grad=True)
+
+
+class HMC(TracePosterior):
+    def __init__(self,
+                 model,
+                 step_size=0.5,
+                 num_steps=3,
+                 warmup_steps=20,
+                 num_samples=1000):
+        self.model = model
+        self.step_size = step_size
+        self.num_steps = num_steps
+        self.num_samples = num_samples
+        self.warmup_steps = warmup_steps
+        # simulation run attributes - will be set in self._setup
+        # at start of run
+        self._cached_param_grads = {}
+        self._args = None
+        self._kwargs = None
+        self._accept_cnt = None
+        self._prototype_trace = None
+        super(HMC, self).__init__()
+
+    def _log_prob(self, z):
+        """
+        Return log pdf of the model with sample sites replayed from z_trace
+        """
+        z_trace = self._prototype_trace.copy()
+        for name, value in z.items():
+            z_trace.nodes[name]['value'] = value
+        model_trace = poutine.trace(poutine.replay(self.model, trace=z_trace)) \
+            .get_trace(*self._args, **self._kwargs)
+        return model_trace.log_pdf()
+
+    def _grad_potential(self, z):
+        log_joint_prob = self._log_prob(z)
+        log_joint_prob.backward()
+        grad_potential = {}
+        for name, value in z.items():
+            grad_potential[name] = -value.grad.clone().detach()
+            grad_potential[name].volatile = False
+        return grad_potential
+
+    def _energy(self, z, r):
+        kinetic_energy = 0.5 * torch.sum(torch.stack([r[name] ** 2 for name in r]))
+        potential_energy = - self._log_prob(z)
+        return kinetic_energy + potential_energy
+
+    def _setup(self, *args, **kwargs):
+        self._accept_cnt = 0
+        self._args = args
+        self._kwargs = kwargs
+        # set the trace prototype to inter-convert between trace object
+        # and dict object used by the integrator
+        self._prototype_trace = poutine.trace(self.model).get_trace(*args, **kwargs)
```

Comment:
Ditto: `detach_grads()` of params during `replay()`.

Reply:
I agree; you have me on board. I will make the change to `ReplayPoutine`.

Reply:
I may be too conservative here, but my inclination has been to have multiple different use cases and unit tests before adding new behavior to `trace` or `replay`, since they're used in almost every algorithm and could therefore absorb any given feature.  Poutines are only useful if we don't have to change them or write new ones for each new algorithm.

This particular behavior will undoubtedly show up in other gradient-based algorithms like SGLD, but my vote would be to wait until we implement those algorithms to incorporate it into `replay`.

Reply:
Makes sense. I will just make a note of it (and create an issue), which we can address after gathering some data points.